lib/screens/v2/widget_instance/pre_fare_line_map.ex

```diff
@@ -0,0 +1,51 @@
+defmodule Screens.V2.WidgetInstance.PreFareLineMap do
+  @moduledoc false
+
+  alias Screens.Config.Screen
+  alias Screens.V2.WidgetInstance
+
+  defstruct screen: nil,
+            asset_url: nil,
+            slot_names: nil,
+            priority: nil
+
+  @type t :: %__MODULE__{
+          screen: Screen.t(),
+          asset_url: String.t(),
+          slot_names: list(WidgetInstance.slot_id()),
```

Comment:
If you are going to define `slot_names` and `priority` directly in here (which I think is fine for this widget), they do not need to be attributes on the struct.

Reply:
Gotcha. I'm actually curious about this one. So I was doing some testing of the instances returned in the candidate generator and I noticed that if I did something like `instance.priority()` it would return `nil` and not `[2]` like I was expecting. But if I call the function at the module level, like `Screens.V2.WidgetInstance.FullLineMap.priority()` it would return that default value.

Why exactly is that the case? I figured it might have to do with functional programming principles and pure functions, but not sure. And also, how does this play into the framework too? In other words, when exactly would those directly defined values/defaults for priority and slot_names get called?

Reply:
With these `WidgetInstance` protocol functions, the instance is not the object that possesses the functions. It is the parameter that you feed into `Screens.V2.WidgetInstance.priority/1`. So in the case where you have separated the `defimpl` functions from the `full_line_map` functions, you either call `FullLineMap.priority/1` to get the specific return for `FullLineMap`, or call `WidgetInstance.priority/1` with any `WidgetInstance` type as a parameter to get the priority value for that particular instance. You can see this used in `screen_data` in `select_best_placement` on line 178. We get the priority in a generic way using `WidgetInstance.priority(instance)`.

Hopefully this makes sense. If not, feel free to reach out.